### PR TITLE
[bindgen] Global `Realm`

### PR DESCRIPTION
**Pull request by kraenhansen**
## What, How & Why?

This will export the `Realm` constructor on the global, as [we do with our old SDK](https://github.com/realm/realm-js/blob/v11/src/js_realm.hpp#L598). This is completely against our goal of eventually [stop polluting the global object](https://github.com/realm/realm-js/issues/2126), but it'll make it easier for existing users to migrate.

I've added a `console.warn` that happens the first time a user access the `Realm` off the global, in hopes that this will help them migrate sooner.

### CODE REVIEW

File: packages/realm/src/safeGlobalThis.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore allow `global`
+  check(typeof global == "object" && global) ||
+  // eslint-disable-next-line no-new-func -- fallback
+  (function () {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore allow `this`
+    return this;
+  })() ||
+  Function("return this")();
```
Comment (kraenhansen) [inline packages/realm/src/safeGlobalThis.ts:47]: We had to introduce this on the `master` branch to work around a bug on older iOS devices with JSC lacking a `globalThis`.

Comment (kraenhansen) [reply packages/realm/src/safeGlobalThis.ts:47]: I copied this from https://github.com/realm/realm-js/blob/v11.4.0/packages/realm-common/src/safeGlobalThis.ts

File: packages/realm/src/safeGlobalThis.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const check = function (it: any) {
+  return it && it.Math == Math && it;
+};
+
+/** @internal */
+// eslint-disable-next-line no-restricted-globals
+export const safeGlobalThis: typeof globalThis & Record<string, unknown> =
+  // eslint-disable-next-line no-restricted-globals
+  check(typeof globalThis == "object" && globalThis) ||
```
Comment (elle-j) [inline packages/realm/src/safeGlobalThis.ts:31]: Could update the `==` in this file to `===` 👍

Comment (kraenhansen) [reply packages/realm/src/safeGlobalThis.ts:31]:
```suggestion
  check(typeof globalThis === "object" && globalThis) ||
```

Comment (elle-j) [reply packages/realm/src/safeGlobalThis.ts]: ps. there were more occurrences

**Review by RedBeard0531 — CHANGES_REQUESTED** [review]

File: packages/realm/src/index.ts [hunk 3]
```diff
@@ -119,10 +119,35 @@ export {
...
+// Patch the global at runtime
+let warnedAboutGlobalRealmUse = false;
+Object.defineProperty(safeGlobalThis, "Realm", {
+  get() {
+    if (!warnedAboutGlobalRealmUse) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        "Your app is relying on a Realm global, which will be removed in realm-js v13,\n",
+        "please update your code to ensure you import Realm via a named ESM-style import:\n\n",
+        'import { Realm } from "realm";\n',
```
[the remark this answers is not in the document]
Comment (kraenhansen) [reply packages/realm/src/index.ts:146]: I added an example for commonjs.

File: packages/realm/src/index.ts [hunk 4]
```diff
@@ -119,10 +119,35 @@ export {
...
 export type ObjectType = string | RealmObjectConstructor;
 
 // Exporting default for backwards compatibility
 export default Realm;
+
+// Exporting a deprecated global for backwards compatibility
+const RealmConstructor = Realm;
+declare global {
+  /** @deprecated Will be removed in v13.0.0. Please use named imports */
+  export type Realm = typeof RealmConstructor;
```
[the remark this answers is not in the document]
Comment (kraenhansen) [reply packages/realm/src/index.ts:134]: I updated this 👍 Thanks.

File: packages/realm/src/safeGlobalThis.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  check(typeof self === "object" && self) ||
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore allow `global`
+  check(typeof global === "object" && global) ||
+  // eslint-disable-next-line no-new-func -- fallback
+  (function () {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore allow `this`
+    return this;
+  })() ||
```
[the remark this answers is not in the document]
Comment (kraenhansen) [reply packages/realm/src/safeGlobalThis.ts:46]: I think that's okay - if we end up here, we're not on a supported runtime anyway.

**Pull request merged**